### PR TITLE
[#5250,#5317] Introduce FindLibCXX cmake module and conditionally define _LIBCPP_DISABLE_DEPRECATION_WARNINGS (master)

### PR DESCRIPTION
**Pull request by SwooshyCueb**
In my work on identifying the best possible solution/workaround for #5317, I came to the conclusion that due to the impact of the scope of the available workarounds, it would be best to only apply them conditionally, only if the `libc++` version being used is 11+.
As part of the changes for #5250, I was already planning on writing a cmake find module for libc++ which would also identify the version of the library, so some (most) of the commits in this repository are actually in service of #5250.
Also, it seemed appropriate to make the compatibility flags for Clang 10 conditional as well, so I did that.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -97,6 +99,37 @@ if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND CMAKE_CXX_COMPILER_VERSION VERSIO
...
+  # existing code, libc++ 11 causes warnings for references to
+  # std::allocator<void>. Maybe a future release of libc++ will fix this.
+  # Should we ever switch to C++20 (or newer), std::allocator<void> problem
+  # will go away, as the explicit template specialization is removed in C++20.
+  #
+  # In the meantime, we have a couple of options for dealing with this.
+  # 1. Define the preprocessor variable _LIBCPP_DISABLE_DEPRECATION_WARNINGS.
+  #    This will silence all deprecation warnings from libc++. References to
+  #    deprecated declarations from other sources will still cause a warning
+  #    and hault the build. This solution/workaround could potentially silence
```
Comment (trel) [inline CMakeLists.txt:119]: hault -> halt

File: CMakeLists.txt [hunk 2]
```diff
@@ -97,6 +99,37 @@ if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND CMAKE_CXX_COMPILER_VERSION VERSIO
...
+  #    valid warnings that would otherwise come from new code that *we*
+  #    introduce.
+  # 2. Pass the compiler option -Wno-error=deprecated-declarations.
+  #    This will prevent references to deprecated declarations from halting
+  #    the build. Warnings will not be silenced, but references to deprecated
+  #    declarations, even those outside of libc++, will not hault the build.
```
Comment (trel) [inline CMakeLists.txt:125]: hault -> halt

File: CMakeLists.txt [hunk 3]
```diff
@@ -97,6 +99,37 @@ if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND CMAKE_CXX_COMPILER_VERSION VERSIO
...
+  # Both of these options have their downsides, but given that option 2 is
+  # much greater in scope, affecting more than just libc++, option 1 will bez
```
Comment (trel) [inline CMakeLists.txt:127]: bez -> be

**Review by trel — COMMENTED** [review]
typo - "compaitbility flags" in fourth commit message

File: CMakeLists.txt (hunk 1 above)
Comment (SwooshyCueb) [reply CMakeLists.txt:119]: I have just now realized that I have probably been spelling this word wrong my entire life.

Comment (trel) [reply CMakeLists.txt]: amazing.

**Review by jasoncoposky — APPROVED** [review]
I've already seen Terrell's comments.

**Comment by trel** [comment]
shall we squash before #?   fine either way.

**Review by korydraughn — APPROVED** [review]
Everything looks fine.

@SwooshyCueb If I build my own version of libc++, what options would I have to set to use it?

@trel @jasoncoposky Why do we need a separate clang runtime dep when clang already comes with all the necessary libs?

**Comment by SwooshyCueb** [comment]
> @&zwnj;SwooshyCueb If I build my own version of libc++, what options would I have to set to use it?

At present, the easiest way is to set `IRODS_EXTERNALS_FULLPATH_CLANG_RUNTIME` to the installation prefix of your libc++. You can do this at configure time by passing `-DIRODS_EXTERNALS_FULLPATH_CLANG_RUNTIME=[prefix]` to cmake.
The `FindLibCXX` module I've added adds the ability to instead specify `LIBCXX_LIBRARY`, `LIBCXX_INCLUDE_DIR`, and (optionally) `LIBCXX_LIBCXXABI_LIBRARY`, but at present this only works for the main irods repository.
 
> @&zwnj;trel @&zwnj;jasoncoposky Why do we need a separate clang runtime dep when clang already comes with all the necessary libs?

`IRODS_EXTERNALS_FULLPATH_CLANG_RUNTIME` is a bit of a misnomer, as is the package name `irods-externals-clang-runtime`. What we are calling the "clang runtime" is actually libc++ and libc++abi, which, depending on how it was built, may not be required for clang at all. Furthermore, GCC and ICC both support using libc++instead of libstdc++, just as clang does.

**Comment by SwooshyCueb** [comment]
> shall we squash before #? fine either way.

I'm gonna vote no on that. If everything else is fine, I'll go ahead and \#.

**Comment by trel** [comment]
seems good - # it up

**Comment by SwooshyCueb** [comment]
\#'d

**Pull request merged**